src/sys/windows/afd.rs

```diff
@@ -0,0 +1,184 @@
+
+#[derive(Debug)]
+pub struct Afd {
+    fd: File,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct AfdPollHandleInfo {
+    pub handle: HANDLE,
+    pub events: ULONG,
+    pub status: NTSTATUS,
+}
+
+unsafe impl Send for AfdPollHandleInfo {}
+unsafe impl Sync for AfdPollHandleInfo {}
+
+#[repr(C)]
+pub struct AfdPollInfo {
+    pub timeout: LARGE_INTEGER,
+    // Can have only value 1.
+    pub number_of_handles: ULONG,
+    pub exclusive: ULONG,
+    pub handles: [AfdPollHandleInfo; 1],
+}
+
+impl fmt::Debug for AfdPollInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AfdPollInfo").finish()
+    }
+}
+
+impl Afd {
+    pub fn new(cp: &CompletionPort) -> io::Result<Afd> {
+        let mut afd_helper_name = OsStr::new(AFD_HELPER_NAME)
+            .encode_wide()
+            .collect::<Vec<_>>();
+
+        let mut afd_helper_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let mut iosb: IO_STATUS_BLOCK = unsafe { std::mem::zeroed() };
+
+        unsafe {
+            let mut objname = UNICODE_STRING {
+                // Lengths are calced in bytes
+                Length: (afd_helper_name.len() * 2) as u16,
+                MaximumLength: (afd_helper_name.len() * 2) as u16,
+                Buffer: afd_helper_name.as_mut_ptr(),
+            };
+            let mut afd_helper_attributes = OBJECT_ATTRIBUTES {
+                Length: size_of::<OBJECT_ATTRIBUTES>() as ULONG,
+                RootDirectory: null_mut() as HANDLE,
+                ObjectName: &mut objname,
+                Attributes: 0 as ULONG,
+                SecurityDescriptor: null_mut() as PVOID,
+                SecurityQualityOfService: null_mut() as PVOID,
+            };
+            let status = NtCreateFile(
+                &mut afd_helper_handle,
+                SYNCHRONIZE,
+                &mut afd_helper_attributes,
+                &mut iosb,
+                null_mut(),
+                0 as ULONG,
+                FILE_SHARE_READ | FILE_SHARE_WRITE,
+                FILE_OPEN,
+                0 as ULONG,
+                null_mut(),
+                0 as ULONG,
+            );
+            if status != STATUS_SUCCESS {
+                return Err(io::Error::from_raw_os_error(
+                    RtlNtStatusToDosError(status) as i32
+                ));
+            }
+            let fd = File::from_raw_handle(afd_helper_handle as RawHandle);
+            let afd = Afd { fd };
+            let token = NEXT_TOKEN.fetch_add(1, Ordering::Relaxed) + 1;
+            cp.add_handle(token, &afd.fd)?;
+            match SetFileCompletionNotificationModes(
+                afd_helper_handle,
+                FILE_SKIP_SET_EVENT_ON_HANDLE,
+            ) {
+                0 => Err(io::Error::last_os_error()),
+                _ => Ok(afd),
+            }
+        }
+    }
+
+    pub fn poll(
+        &self,
+        info: &mut AfdPollInfo,
+        iosb: &mut IO_STATUS_BLOCK,
+        apccontext: PVOID,
```

Comment:
Could we get some comments on the `apccontext` argument here?

Is this passing ownership of the data into the function? If it is a reference, what is the lifetime of the pointer?

cc @piscisaureus

Reply:
As I observed from other snippets in Google, Apccontext is probably passed directly to CompletionStatus as Overlapped value.

Reply:
It looks like the only place this is called is in `selector.rs`. Instead of passing in the raw pointer, we could pass in an `&Arc<SockState>` and do the `clone` -> `into_raw` bit in here.

Reply:
I thought it would be general case as Afd copied only pointer not using the internal data.

Reply:
See https://github.com/tokio-rs/mio/pull/1034#discussion_r304438939